src/iceberg/table_scan.h

```diff
@@ -0,0 +1,204 @@
+#include "iceberg/manifest_entry.h"
+#include "iceberg/type_fwd.h"
+
+namespace iceberg {
+
+class ICEBERG_EXPORT ScanTask {
+ public:
+  virtual ~ScanTask() = default;
+
+  /// \brief The number of bytes that should be read by this scan task.
+  virtual int64_t size_bytes() const = 0;
+
+  /// \brief The number of files that should be read by this scan task.
+  virtual int32_t files_count() const = 0;
+
+  /// \brief The number of rows that should be read by this scan task.
+  virtual int64_t estimated_row_count() const = 0;
+};
+
+/// \brief Represents a task to scan a portion of a data file.
+class ICEBERG_EXPORT FileScanTask : public ScanTask {
+ public:
+  FileScanTask(std::shared_ptr<DataFile> file,
+               std::vector<std::shared_ptr<DataFile>> delete_files, int64_t start,
+               int64_t length, std::shared_ptr<Expression> residual);
+
+  /// \brief The data file that should be read by this scan task.
+  const std::shared_ptr<DataFile>& data_file() const;
+
+  /// \brief The delete files that should be read by this scan task.
+  const std::vector<std::shared_ptr<DataFile>>& delete_files() const;
+
+  /// \brief The byte offset in the data file where the scan should start.
+  int64_t start() const;
+
+  /// \brief The length in bytes to scan from the start offset.
+  int64_t length() const;
+
+  /// \brief The residual expression to apply after scanning the data file.
+  const std::shared_ptr<Expression>& residual() const;
+
+  int64_t size_bytes() const override;
+  int32_t files_count() const override;
+  int64_t estimated_row_count() const override;
+
+ private:
+  std::shared_ptr<DataFile> data_file_;                  ///< Data file metadata.
+  std::vector<std::shared_ptr<DataFile>> delete_files_;  ///< Delete files metadata.
+
+  int64_t start_;   ///< Start byte offset.
+  int64_t length_;  ///< Length in bytes to scan.
+
+  std::shared_ptr<Expression> residual_;  ///< Residual expression to apply.
+};
+
+/// \brief Scan context holding snapshot and scan-specific metadata.
+struct TableScanContext {
+  std::shared_ptr<TableMetadata> table_metadata;  ///< Table metadata.
+  std::shared_ptr<Snapshot> snapshot;             ///< Snapshot to scan.
+  std::shared_ptr<Schema> projected_schema;       ///< Projected schema.
+  std::shared_ptr<Expression> filter;             ///< Filter expression to apply.
+  bool case_sensitive = false;                    ///< Whether the scan is case-sensitive.
+  std::unordered_map<std::string, std::string>
+      options;                   ///< Additional options for the scan.
+  std::optional<int64_t> limit;  ///< Optional limit on the number of rows to scan.
+};
+
+/// \brief Builder class for creating TableScan instances.
+class ICEBERG_EXPORT TableScanBuilder {
+ public:
+  /// \brief Constructs a TableScanBuilder for the given table.
+  /// \param table The table to scan.
+  /// \param table_metadata The metadata of the table to scan.
+  explicit TableScanBuilder(const Table& table,
```

Comment:
Why do we need both `table` and `table_metadata`?

Reply:
`Table` is just a wrapper around `TableMetadata`. I think we could either (1) expose `TableMetadata` via `Table` (and thus remove `table_metadata` from ctor), or (2) remove `table` and add `file_io` to ctor. I prefer (2) here.